### ISSUE TITLE
M1 CI is generally flaky with SIGSEGVs, exit code 134s, and flaky tests

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#84022]
## TL;DR
1. Why does MPS segfault so much?
2. I suspect some of the flakiness is related which is why this is one issue but if they're separate, I can make more issues.

In the past day, 5+ out of 100 commits (5%) failed on trunk due to MPS flakiness. https://hud.pytorch.org/minihud/pytorch/pytorch/master/1?per_page=50&name_filter=trunk%20%2F%20macos-12-py3-arm64%20%2F%20test%20(default%2C%201%2C%202%2C%20macos-m1-12)

This is past our requirement of 2.5%  flakiness for trunk jobs. This might just be a flaky week for M1 machines, but anecdotally, mps has been consistently flaky and as we gather longer term stats, I figured we should start tackling some of these recurring issues:

## SIGSEGVs
These seem to appear near test_output_match_split_cpu tests.
https://github.com/pytorch/pytorch/runs/7985869067?check_suite_focus=true
```
2022-08-24T01:34:15.5890920Z   test_output_match_split_cpu_float32 (__main__.TestConsistencyCPU) ... skip: Test is disabled because an issue exists disabling it: https://github.com/pytorch/pytorch/issues/82961 for platform(s) mac, macos. If you're seeing this on your local machine and would like to enable this test, please make sure CI is not set and you are not using the flag --import-disabled-tests. (0.001s)
2022-08-24T01:34:15.5953050Z   test_output_match_split_cpu_int16 (__main__.TestConsistencyCPU) ... ok (0.006s)
2022-08-24T01:34:15.6011260Z   test_output_match_split_cpu_int32 (__main__.TestConsistencyCPU) ... Traceback (most recent call last):
2022-08-24T01:34:15.6011680Z   File "/Users/ec2-user/runner/_work/pytorch/pytorch/test/run_test.py", line 1065, in <module>
2022-08-24T01:34:15.6011860Z     main()
2022-08-24T01:34:15.6012090Z   File "/Users/ec2-user/runner/_work/pytorch/pytorch/test/run_test.py", line 1043, in main
2022-08-24T01:34:15.6012310Z     raise RuntimeError(err_message)
2022-08-24T01:34:15.6012500Z RuntimeError: test_mps failed! Received signal: SIGSEGV
2022-08-24T01:34:15.7141600Z ##[error]Process completed with exit code 1.
```

## Exit code 134
https://github.com/pytorch/pytorch/runs/7986589182?check_suite_focus=true
https://github.com/pytorch/pytorch/runs/7982119217?check_suite_focus=true
https://github.com/pytorch/pytorch/runs/7962486586?check_suite_focus=true 
https://github.com/pytorch/pytorch/runs/8035232454?check_suite_focus=true
```
2022-08-24T02:57:39.5758280Z + cmake ..
2022-08-24T02:57:39.7744960Z dyld[45722]: Library not loaded: @rpath/libzstd.1.dylib
2022-08-24T02:57:39.7747200Z   Referenced from: /opt/homebrew/Caskroom/miniconda/base/pkgs/cmake-3.22.1-hae769c0_0/bin/cmake
2022-08-24T02:57:39.7754840Z   Reason: tried: '/var/folders/nz/j6p8yfhx1mv_0grj5xl4650h0000gp/T/croot-myzr18j9/cmake_1641978514075/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/libzstd.1.dylib' (no such file), '/opt/homebrew/Caskroom/miniconda/base/pkgs/cmake-3.22.1-hae769c0_0/bin/../lib/libzstd.1.dylib' (no such file), '/var/folders/nz/j6p8yfhx1mv_0grj5xl4650h0000gp/T/croot-myzr18j9/cmake_1641978514075/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/libzstd.1.dylib' (no such file), '/opt/homebrew/Caskroom/miniconda/base/pkgs/cmake-3.22.1-hae769c0_0/bin/../lib/libzstd.1.dylib' (no such file), '/usr/local/lib/libzstd.1.dylib' (no such file), '/usr/lib/libzstd.1.dylib' (no such file)
2022-08-24T02:57:39.7760630Z .jenkins/pytorch/macos-test.sh: line 134: 45722 Abort trap: 6           CMAKE_PREFIX_PATH="$SITE_PACKAGES/torch" cmake ..
2022-08-24T02:57:39.7794100Z ##[error]Process completed with exit code 134.
```

## Other flaky tests
* Created an issue for test_proxy_tensor https://github.com/pytorch/pytorch/issues/84019 

### Versions

CI, master

cc @ezyang @gchanan @zou3519 @seemethere @malfet @pytorch/pytorch-dev-infra @kulinseth @albanD

**Comment by seemethere** [comment]
NOTE: The runners these run on are _pet_ instances meaning that they do not get cleaned up which might also make it so that there is a high possibility that there are _bad_ nodes

**Comment by janeyx99** [comment]
@seemethere Is the advisory there to report the bad nodes to yall to take them offline?

**Comment by janeyx99** [comment]
i-0a344d9728d8321e6 seems like a bad node then (common in two of the above links).

**Comment by malfet** [comment]
Edited the title - it's not MPS CI that is flaky(see this link - not a single failure in last 150 runs: https://hud.pytorch.org/hud/pytorch/pytorch/master/3?per_page=50&name_filter=Run%20MPS ), but a general M1 CI

**Comment by seemethere** [comment]
SIGSEGV usually indicates a segmentation fault, are we sure we’re not hitting some rare edge case in this scenario? Like maybe this is a real bug

**Comment by seemethere** [comment]
Exit code 134 is a bit of a red herring in this case. The real error is `Abort trap: 6` which [appears to indicate](https://stackoverflow.com/questions/35247696/why-is-my-program-giving-me-an-abort-trap-6-error) an issue with the thing that is being compiled. I'm not entirely sure how we get into these failure states though and why it's only happening infrequently.

**Comment by albanD** [comment]
Well, the first failure is indeed a segfault while running the mps tests.

**Comment by janeyx99** [comment]
**As a reminder, per [the current CI rules](https://docs.google.com/document/d/1Mxrjrs6kr7oBukmwSjG4LkfHpbAEw-bp2T-oxpzeDzg/edit#), jobs that have flakiness over 2.5% will be removed from core CI.** The M1 CI would qualify if it's consistently flaky over 2.5% of the time. Since its inception, M1 CI has been red ~10% of ~2k workflow job runs. The DevX team is committed to keeping flakiness below 2.5% so that users can trust our CI. M1 flakiness currently impacts every commit on trunk as well as land checks. For example, users have been blocked on landing their PRs because M1 tests failed flakily during land checks.

This would mean that we will no longer validate and track that every commit passes M1 testing and we will not revert commits that break M1. We can still run the tests on a periodic cadence, given machine availability, and M1 engineers can use that information to forward fix brokenness.

**Comment by janeyx99** [comment]
Related: https://github.com/pytorch/pytorch/issues/84196

**Comment by DenisVieriu97** [comment]
Hi @janeyx99! Thank you for the report.
I have tried reproducing this issue, but with no success so far unfortunately. Could you please let me know if this is related to some specific builds, or is it happening randomly? 
I have tried with a local build from https://github.com/kulinseth/pytorch (mps_master branch), using the following flags:
```
CC=`which clang` CXX=`which clang++` DEBUG=0 BUILD_CAFFE2_OPS=0 USE_DISTRIBUTED=0 USE_MKLDNN=0 USE_CUDA=0 BUILD_TEST=1 USE_FBGEMM=0 USE_NNPACK=0 USE_QNNPACK=0 python setup.py build --cmake --cmake-only

CC=`which clang` CXX=`which clang++` DEBUG=0 BUILD_CAFFE2_OPS=0 USE_DISTRIBUTED=0 USE_MKLDNN=0 USE_CUDA=0 BUILD_TEST=1 USE_FBGEMM=0 USE_NNPACK=0 USE_QNNPACK=0 WERROR=1 python setup.py install
``` 
If you spot anything wrong with the flags, please let me know. I will try in the meantime to reproduce it with a nightly build.

**Comment by janeyx99** [comment]
Hi @DenisVieriu97, thanks for looking into the build! Using the nightly is a good next step. That said, I've checked the last few days of logs and I haven't seen the exit 134 error. 

However, the SIGSEGVs during tests are still prevalent and are not isolated to only one group of tests and may be easier to reproduce. For example: https://github.com/pytorch/pytorch/actions/runs/3039864026/jobs/4895817665. It may be good to prioritize why the SIGSEGVs are so common as things are building 😛

**Comment by malfet** [comment]
I believe I've mitigated the issue by not running MPS tests during regular M1 testing, see https://github.com/pytorch/pytorch/pull/85500
Another problem that might have contributed to flakiness were disk space issues that were addressed by @huydhn 
Last potential source of flakiness are cpp_extensions tests when we compile and load .dylib files, which might occasionally trigger M1 digital signature validation failures

**Comment by albanD** [comment]
Thanks for the update!
Should we close this issue as mitigated for now?
